File: python/pyautd3/modulation/fir.py
"""
File: fir.py
Project: modulation
Created Date: 12/10/2023
Author: Shun Suzuki
-----
Last Modified: 13/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


from typing import Generic, TypeVar

from pyautd3.internal.modulation import IModulation
from pyautd3.native_methods.autd3capi import NativeMethods as Base
from pyautd3.native_methods.autd3capi_def import ModulationPtr

M = TypeVar("M", bound=IModulation)


class LPF(IModulation, Generic[M]):
    """Low pass filter."""

    _m: M
    _n_taps: int
    _cutoff: float

    def __init__(self: "LPF", m: M, n_taps: int, cutoff: float) -> None:
        self._m = m
        self._n_taps = n_taps
        self._cutoff = cutoff

    def _modulation_ptr(self: "LPF") -> ModulationPtr:
        return Base().modulation_with_low_pass(self._m._modulation_ptr(), self._n_taps, self._cutoff)


class HPF(IModulation, Generic[M]):
    """High pass filter."""

    _m: M
    _n_taps: int
    _cutoff: float

    def __init__(self: "HPF", m: M, n_taps: int, cutoff: float) -> None:
        self._m = m
        self._n_taps = n_taps
        self._cutoff = cutoff

    def _modulation_ptr(self: "HPF") -> ModulationPtr:
        return Base().modulation_with_high_pass(self._m._modulation_ptr(), self._n_taps, self._cutoff)


class BPF(IModulation, Generic[M]):
    """Band pass filter."""

    _m: M
    _n_taps: int
    _f_low: float
    _f_high: float

    def __init__(self: "BPF", m: M, n_taps: int, f_low: float, f_high: float) -> None:
        self._m = m
        self._n_taps = n_taps
        self._f_low = f_low
        self._f_high = f_high

    def _modulation_ptr(self: "BPF") -> ModulationPtr:
        return Base().modulation_with_band_pass(self._m._modulation_ptr(), self._n_taps, self._f_low, self._f_high)


class BSF(IModulation, Generic[M]):
    """Band stop filter."""

    _m: M
    _n_taps: int
    _f_low: float
    _f_high: float

    def __init__(self: "BSF", m: M, n_taps: int, f_low: float, f_high: float) -> None:
        self._m = m
        self._n_taps = n_taps
        self._f_low = f_low
        self._f_high = f_high

    def _modulation_ptr(self: "BSF") -> ModulationPtr:
        return Base().modulation_with_band_stop(self._m._modulation_ptr(), self._n_taps, self._f_low, self._f_high)


def __with_low_pass(self: M, n_taps: int, cutoff: float) -> LPF:
    return LPF(self, n_taps, cutoff)


def __with_high_pass(self: M, n_taps: int, cutoff: float) -> HPF:
    return HPF(self, n_taps, cutoff)


def __with_band_pass(self: M, n_taps: int, f_low: float, f_high: float) -> BPF:
    return BPF(self, n_taps, f_low, f_high)


def __with_band_stop(self: M, n_taps: int, f_low: float, f_high: float) -> BSF:
    return BSF(self, n_taps, f_low, f_high)


IModulation.with_low_pass = __with_low_pass  # type: ignore[method-assign]
IModulation.with_high_pass = __with_high_pass  # type: ignore[method-assign]
IModulation.with_band_pass = __with_band_pass  # type: ignore[method-assign]
IModulation.with_band_stop = __with_band_stop  # type: ignore[method-assign]

File: dotnet/cs/src/Modulation/FIR.cs
/*
 * File: FIR.cs
 * Project: Modulation
 * Created Date: 12/10/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 07/11/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */


#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
using Math = UnityEngine.Mathf;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp.Modulation
{
    public sealed class Lpf : Internal.Modulation
    {
        private readonly Internal.Modulation _m;
        private readonly uint _nTaps;
        private readonly float_t _cutoff;

        public Lpf(Internal.Modulation m, uint nTaps, float_t cutoff)
        {
            _m = m;
            _nTaps = nTaps;
            _cutoff = cutoff;
        }

        internal override ModulationPtr ModulationPtr()
        {
            return NativeMethodsBase.AUTDModulationWithLowPass(_m.ModulationPtr(), _nTaps, _cutoff);
        }
    }


    public sealed class Hpf : Internal.Modulation
    {
        private readonly Internal.Modulation _m;
        private readonly uint _nTaps;
        private readonly float_t _cutoff;

        public Hpf(Internal.Modulation m, uint nTaps, float_t cutoff)
        {
            _m = m;
            _nTaps = nTaps;
            _cutoff = cutoff;
        }

        internal override ModulationPtr ModulationPtr()
        {
            return NativeMethodsBase.AUTDModulationWithHighPass(_m.ModulationPtr(), _nTaps, _cutoff);
        }
    }

    public sealed class Bpf : Internal.Modulation
    {
        private readonly Internal.Modulation _m;
        private readonly uint _nTaps;
        private readonly float_t _fLow;
        private readonly float_t _fHigh;

        public Bpf(Internal.Modulation m, uint nTaps, float_t fLow, float_t fHigh)
        {
            _m = m;
            _nTaps = nTaps;
            _fLow = fLow;
            _fHigh = fHigh;
        }

        internal override ModulationPtr ModulationPtr()
        {
            return NativeMethodsBase.AUTDModulationWithBandPass(_m.ModulationPtr(), _nTaps, _fLow, _fHigh);
        }
    }

    public sealed class Bsf : Internal.Modulation
    {
        private readonly Internal.Modulation _m;
        private readonly uint _nTaps;
        private readonly float_t _fLow;
        private readonly float_t _fHigh;

        public Bsf(Internal.Modulation m, uint nTaps, float_t fLow, float_t fHigh)
        {
            _m = m;
            _nTaps = nTaps;
            _fLow = fLow;
            _fHigh = fHigh;
        }

        internal override ModulationPtr ModulationPtr()
        {
            return NativeMethodsBase.AUTDModulationWithBandStop(_m.ModulationPtr(), _nTaps, _fLow, _fHigh);
        }
    }

    public static class FirModulationExtensions
    {
        public static Lpf WithLowPass(this Internal.Modulation s, uint nTaps, float_t cutoff)
        {
            return new Lpf(s, nTaps, cutoff);
        }

        public static Hpf WithHighPass(this Internal.Modulation s, uint nTaps, float_t cutoff)
        {
            return new Hpf(s, nTaps, cutoff);
        }

        public static Bpf WithBandPass(this Internal.Modulation s, uint nTaps, float_t fLow, float_t fHigh)
        {
            return new Bpf(s, nTaps, fLow, fHigh);
        }

        public static Bsf WithBandStop(this Internal.Modulation s, uint nTaps, float_t fLow, float_t fHigh)
        {
            return new Bsf(s, nTaps, fLow, fHigh);
        }
    }
}
